File: ios/Reguerta/Reguerta/Data/Commitments/ChainedSeasonalCommitmentRepository.swift
import Foundation

actor ChainedSeasonalCommitmentRepository<
    Primary: SeasonalCommitmentRepository,
    Fallback: SeasonalCommitmentRepository
>: SeasonalCommitmentRepository {
    private let primary: Primary
    private let fallback: Fallback

    init(primary: Primary, fallback: Fallback) {
        self.primary = primary
        self.fallback = fallback
    }

    func activeCommitments(userId: String) async throws -> [SeasonalCommitment] {
        let primaryItems = try await primary.activeCommitments(userId: userId)
        if !primaryItems.isEmpty {
            return primaryItems
        }
        return try await fallback.activeCommitments(userId: userId)
    }
}

File: android/Reguerta/app/src/main/java/com/reguerta/user/data/commitments/ChainedSeasonalCommitmentRepository.kt
package com.reguerta.user.data.commitments

import com.reguerta.user.domain.commitments.SeasonalCommitment
import com.reguerta.user.domain.commitments.SeasonalCommitmentRepository

class ChainedSeasonalCommitmentRepository(
    private val primary: SeasonalCommitmentRepository,
    private val fallback: SeasonalCommitmentRepository,
) : SeasonalCommitmentRepository {
    override suspend fun getActiveCommitmentsForUser(userId: String): List<SeasonalCommitment> {
        val primaryCommitments = primary.getActiveCommitmentsForUser(userId)
        return if (primaryCommitments.isNotEmpty()) primaryCommitments else fallback.getActiveCommitmentsForUser(userId)
    }
}
